netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```diff
@@ -97,24 +113,79 @@
   private Status goAwayStatus;
   private int nextStreamId;
 
-  public NettyClientHandler(BufferingHttp2ConnectionEncoder encoder, Http2Connection connection,
-                            Http2FrameReader frameReader,
-                            int flowControlWindow) {
-    this(encoder, connection, frameReader, flowControlWindow, Ticker.systemTicker());
+  static NettyClientHandler newHandler(ClientTransport.Listener listener,
+                                       int flowControlWindow, int maxHeaderListSize,
+                                       Ticker ticker) {
+    Preconditions.checkArgument(maxHeaderListSize > 0, "maxHeaderListSize must be positive");
+    Http2HeadersDecoder headersDecoder =
+        new DefaultHttp2HeadersDecoder(maxHeaderListSize, Http2CodecUtil.DEFAULT_HEADER_TABLE_SIZE);
+    Http2FrameReader frameReader = new DefaultHttp2FrameReader(headersDecoder);
+    Http2FrameWriter frameWriter = new DefaultHttp2FrameWriter();
+    Http2Connection connection = new DefaultHttp2Connection(false);
+    return newHandler(connection, frameReader, frameWriter, listener, flowControlWindow, ticker);
   }
 
   @VisibleForTesting
-  NettyClientHandler(BufferingHttp2ConnectionEncoder encoder, Http2Connection connection,
-                     Http2FrameReader frameReader, int flowControlWindow, Ticker ticker) {
-    super(new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader), encoder,
-            createInitialSettings(flowControlWindow));
+  static NettyClientHandler newHandler(Http2Connection connection,
+                                       Http2FrameReader frameReader,
+                                       Http2FrameWriter frameWriter,
+                                       final ClientTransport.Listener listener,
+                                       int flowControlWindow,
+                                       Ticker ticker) {
+    Preconditions.checkNotNull(connection, "connection");
+    Preconditions.checkNotNull(frameReader, "frameReader");
+    Preconditions.checkNotNull(listener, "listener");
+    Preconditions.checkArgument(flowControlWindow > 0, "flowControlWindow must be positive");
+    Preconditions.checkNotNull(ticker, "ticker");
+
+    Http2FrameLogger frameLogger = new Http2FrameLogger(LogLevel.DEBUG, NettyClientHandler.class);
+    frameReader = new Http2InboundFrameLogger(frameReader, frameLogger);
+    frameWriter = new Http2OutboundFrameLogger(frameWriter, frameLogger);
+
+    BufferingHttp2ConnectionEncoder encoder = new BufferingHttp2ConnectionEncoder(
+        new DefaultHttp2ConnectionEncoder(connection, frameWriter)) {
+      private boolean firstSettings = true;
+
+      @Override
+      public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+        if (firstSettings) {
+          firstSettings = false;
+          try {
+            listener.transportReady();
+          } catch (Exception e) {
```

Comment:
I don't think you need the try-catch any more. I think you had it because of Future.get().